sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADClientRegistrationRepository.java

```diff
@@ -3,83 +3,402 @@
 import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
 
-import java.util.ArrayList;
-import java.util.HashMap;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Optional;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.azure.spring.aad.AADAuthorizationGrantType.AUTHORIZATION_CODE;
+import static com.azure.spring.aad.AADConditions.BEARER_TOKEN_AUTHENTICATION_TOKEN_CLASS_NAME;
 
 
 /**
  * Manage all AAD oauth2 clients configured by property "azure.activedirectory.xxx"
  */
-public abstract class AADClientRegistrationRepository implements ClientRegistrationRepository {
+public class AADClientRegistrationRepository
+    implements ClientRegistrationRepository, Iterable<ClientRegistration> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AADClientRegistrationRepository.class);
 
     public static final String AZURE_CLIENT_REGISTRATION_ID = "azure";
 
-    protected final AzureClientRegistration azureClient;
-    protected final List<ClientRegistration> otherClients;
-    protected final Map<String, ClientRegistration> allClients;
+    protected final AzureClientRegistration azureClientRegistration;
+    protected final Map<String, ClientRegistration> delegatedClientRegistrations;
+    protected final Map<String, ClientRegistration> allClientRegistrations;
     protected final AADAuthenticationProperties properties;
+    private final Boolean isWebApiClientMode;
+    private final Boolean isWebAppClientMode;
+    private final Boolean isAllInClientMode;
 
-    public AADClientRegistrationRepository(AzureClientRegistration azureClient,
-                                           List<ClientRegistration> otherClients,
-                                           AADAuthenticationProperties properties) {
-        this.azureClient = azureClient;
-        this.otherClients = new ArrayList<>(otherClients);
+    public AADClientRegistrationRepository(AADAuthenticationProperties properties) {
         this.properties = properties;
+        this.isWebAppClientMode = isWebAppClientMode();
+        this.isWebApiClientMode = isWebApiClientMode();
+        this.isAllInClientMode = isAllInClientMode();
+        properties.getAuthorizationClients().forEach(this::handleDefaultAADAuthorizationGrantType);
+        this.azureClientRegistration = azureClientRegistration();
+        this.delegatedClientRegistrations = delegatedClientRegistrations();
+        this.allClientRegistrations = allClientRegistrations();
+    }
+
+    private AzureClientRegistration azureClientRegistration() {
+        if (!isWebAppClientMode && !isAllInClientMode) {
```

Comment:
if `!isWebAppClientMode && !isAllInClientMode`, maybe we do not need to use `AADClientRegistrationRepository`, we can use InMemoryClientRegistrationRepository directly.

Reply:
The `InMemoryClientRegistrationRepository ` is also unused. In order to make the structure consistent, just created an empty `AzureClientRegistration` if it is only resource server.